test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1307 @@
+		return;
+	}
+
+	if (spi_to_sa_map[dir][spi] != NULL) {
+		ODPH_ERR("Duplicate SA config %u\n", spi);
+		return;
+	}
+
+	odp_queue_param_init(&q_param);
+	q_param.type = ODP_QUEUE_TYPE_SCHED;
+	q_param.sched.prio = odp_schedule_max_prio();
+	q_param.sched.sync = config->mode == 0U ? ODP_SCHED_SYNC_ORDERED : ODP_SCHED_SYNC_PARALLEL;
+	q_param.sched.group = ODP_SCHED_GROUP_ALL;
+	config->ev_qs[config->num_sas] = odp_queue_create(PROG_NAME, &q_param);
+
+	if (config->ev_qs[config->num_sas] == ODP_QUEUE_INVALID) {
+		ODPH_ERR("Error creating SA destination queue for SA %u\n", spi);
+		return;
+	}
+
+	src_ip = odp_cpu_to_be_32(src_ip);
+	dst_ip = odp_cpu_to_be_32(dst_ip);
+	odp_ipsec_sa_param_init(&sa_param);
+	sa_param.proto = ODP_IPSEC_ESP;
+	sa_param.mode = ODP_IPSEC_MODE_TUNNEL;
+	sa_param.spi = spi;
+	sa_param.dest_queue = config->ev_qs[config->num_sas];
+
+	if (dir > 0U) {
+		sa_param.dir = ODP_IPSEC_DIR_OUTBOUND;
+		sa_param.outbound.tunnel.ipv4.src_addr = &src_ip;
+		sa_param.outbound.tunnel.ipv4.dst_addr = &dst_ip;
+	} else {
+		sa_param.dir = ODP_IPSEC_DIR_INBOUND;
+		sa_param.inbound.lookup_mode = ODP_IPSEC_LOOKUP_SPI;
+		sa_param.inbound.antireplay_ws = ar_ws;
+	}
+
+	crypto_param.cipher_alg = cipher_idx;
+	crypto_param.cipher_key.data = cipher_key;
+	crypto_param.cipher_key.length = strlen((const char *)cipher_key);
+	crypto_param.cipher_key_extra.data = cipher_key_extra;
+	crypto_param.cipher_key_extra.length = strlen((const char *)cipher_key_extra);
+	crypto_param.auth_alg = auth_idx;
+	crypto_param.auth_key.data = auth_key;
+	crypto_param.auth_key.length = strlen((const char *)auth_key);
+	crypto_param.auth_key_extra.data = auth_key_extra;
+	crypto_param.auth_key_extra.length = strlen((const char *)auth_key_extra);
+	crypto_param.icv_len = icv_len;
+	sa_param.crypto = crypto_param;
+	sa = odp_ipsec_sa_create(&sa_param);
+
+	if (sa == ODP_IPSEC_SA_INVALID) {
+		(void)odp_queue_destroy(config->ev_qs[config->num_sas]);
+		ODPH_ERR("Error creating SA handle\n");
+		return;
+	}
+
+	config->sas[config->num_sas] = sa;
+	spi_to_sa_map[dir][spi] = &config->sas[config->num_sas];
+	++config->num_sas;
+}
+
+static void parse_sas(prog_config_t *config)
+{
+	odp_ipsec_capability_t ipsec_capa;
+	FILE *file;
+	int cipher_idx, auth_idx;
+	uint32_t dir, spi, icv_len;
+	char src_ip[15U] = { 0 }, dst_ip[15U] = { 0 };
+	uint8_t cipher_key[64U] = { 0U }, cipher_key_extra[4U] = { 0U }, auth_key[64U] = { 0U },
+	auth_key_extra[4U] = { 0U };
+
+	if (odp_ipsec_capability(&ipsec_capa) < 0) {
+		ODPH_ERR("Error querying IPsec capabilities\n");
+		return;
+	}
+
+	if (!setup_ipsec(config))
+		return;
+
+	if (config->sa_conf_file == NULL) {
+		ODPH_ERR("Invalid SA configuration file\n");
+		return;
+	}
+
+	file = fopen(config->sa_conf_file, "r");
+
+	if (file == NULL) {
+		ODPH_ERR("Error opening SA configuration file: %s\n", strerror(errno));
+		return;
+	}
+
+	while (fscanf(file, "%u%u%s%s%d%s%s%d%s%s%u", &dir, &spi, src_ip, dst_ip,
```

Comment:
This is a bit fragile for the string arguments. I think it would be better to include max field width in the string conversions (i.e. %u%u%14s%14s%d%63s%3s%d%63s%3s%u") if using sscanf and the current array sizes.

Reply:
Some trickery is needed if a string ends up being longer than there is space defined in the format as the rest of that string in the file should then be skipped. I'll think this separately and leave this as is for now.